persistence/remote/src/main/java/org/infinispan/persistence/remote/RemoteStore.java

```diff
@@ -319,45 +363,46 @@ private Object getValue(MarshallableEntry entry) {
    }
 
    @Override
-   public CompletionStage<Void> bulkUpdate(Publisher<MarshallableEntry<? extends K, ? extends V>> publisher) {
-      return Flowable.fromPublisher(publisher)
-            .buffer(configuration.maxBatchSize())
-            // TODO: this is blocking - when the full non blocking SPI is introduced this needs to chain returns
-            // of the putAll to ensure that only a single batch is sent at a time
-            .doOnNext(entries -> remoteCache.putAll(entries.stream().collect(Collectors.toMap(this::getKey, this::getValue))))
-            .doOnError(PersistenceException::new)
-            .ignoreElements()
+   public CompletionStage<Void> batch(int publisherCount, Publisher<SegmentedPublisher<Object>> removePublisher,
+         Publisher<SegmentedPublisher<MarshallableEntry<K, V>>> writePublisher) {
+      Completable removeCompletable = Flowable.fromPublisher(removePublisher)
+            .flatMap(sp -> Flowable.fromPublisher(sp), publisherCount)
+            .map(key -> key instanceof WrappedByteArray ? ((WrappedByteArray) key).getBytes() : key)
```

Comment:
We should probably add a static `unwrap` method, as this code is repeated several times. We can then use a method reference here.

Reply:
Okay.